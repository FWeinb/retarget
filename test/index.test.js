import select from '../src/index'

test('basic', () => {
  const selector = select`profile.name.last`

  // function reducer(state, action) {
  //   const lastNameSelector = selector(state);
  //   // getting the value is handy in components and mapStateToProps
  //   let currentVal = lastNameSelector.get();

  //   // But I think its more handy for setting
  //   lastNameSelector.set(action.payload);

  //   // if it was an array value
  //   lastNameSelector.push("foo");
  //   let someIndex = action.payload;
  //   lastNameSelector.remove(someIndex);

  //   return lastNameSelector.save();
  // }

  const arraySelector = select(['profile', 'name', 'first'])
  const STATE = {
    profile: {
      name: {
        first: 'Waylon',
        last: 'Jennings'
      }
    }
  }

  expect(selector(STATE)).toMatchSnapshot()
  expect(arraySelector(STATE)).toMatchSnapshot()
})

test('nested', () => {
  const STATE2 = {
    users: {
      '1': {
        profile: {
          name: {
            first: 'Waylon',
            last: 'Jennings'
          }
        }
      },
      '2': {
        profile: {
          name: {
            first: 'Waylon',
            last: 'Jennings'
          }
        }
      }
    }
  }

  const userSelector = select`users`
  const selector2 = select`${userSelector}.2.profile.name.last`
  expect(selector2(STATE2)).toMatchSnapshot()
})

test('compound', () => {
  const STATE3 = {
    entities: {
      users: {
        '1': {
          profile: {
            name: {
              first: 'Waylon',
              last: 'Jennings'
            }
          }
        },
        '2': {
          profile: {
            name: {
              first: 'Waylon',
              last: 'Jennings'
            }
          }
        }
      }
    }
  }
  // I can't figure out how to dedupe selectors
  // If you use the 2 above one after the other you will get ['profile', 'name', 'last'] twice in the key

  // I think going from right to left with the head of the new key array
  // looking for a match and then overwriting (left to right) from there might work
  // need to do some sort of check before doing this check as it is stupid innefficent
  // unless we know there is a hit
  const selectorA = select`users.1`
  const selectorB = select`profile`
  const selectorC = select`first`
  const compoundSelector = select`entities.${selectorA}${selectorB}.name.${selectorC}`
  expect(compoundSelector(STATE3)).toMatchSnapshot()
})
