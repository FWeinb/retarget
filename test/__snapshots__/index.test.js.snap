// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `"Jennings"`;

exports[`basic 2`] = `"Waylon"`;

exports[`compound 1`] = `"Waylon"`;

exports[`fun fun functions 1`] = `"Jennings"`;

exports[`nested 1`] = `"Jennings"`;
